File: ganapathy_pavers/ganapathy_pavers/report/vehicle_usage/vehicle_usage.js
// Copyright (c) 2023, Thirvusoft and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Vehicle Usage"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_start(),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_end(),
		},
		{
			fieldname: "purpose",
			label: __("Purpose"),
			fieldtype: "MultiSelectList",
			get_data: async function () {
				let res = [], opts = "";
				frappe.model.with_doctype("Vehicle Log", () => {
					frappe.get_meta("Vehicle Log").fields.every(row => {
						if (row.fieldname == "select_purpose") {
							opts = row.options
							return false
						}
						return true
					})
				});

				opts.split('\n').forEach(opt => {
					if (opt) {
						res.push({
							value: opt,
							description: opt
						});
					}
				});

				return res
			}
		}
	]
};

File: ganapathy_pavers/ganapathy_pavers/report/vehicle_usage/vehicle_usage.py
# Copyright (c) 2023, Thirvusoft and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
	columns, data = get_columns(filters), get_data(filters)
	chart_data = {
		"data": {
			"labels": [(i.get('vehicle') or '') for i in data],
			"datasets": [
				{"name": "Distance Travelled", "values": [(i.get('distance') or 0) for i in data]},
			],
		},
		"type": "bar",
	}
	return columns, data, None, chart_data

def get_data(filters):
	return frappe.db.sql(f"""
		SELECT
			vl.license_plate as vehicle,
			SUM(vl.today_odometer_value) as distance
		FROM `tabVehicle Log` vl
		WHERE
			vl.docstatus = 1 AND
		    IFNULL(vl.today_odometer_value, 0) != 0 AND
		    CASE
				WHEN IFNULL('{filters.get('from_date') or ''}', '') != ''
					THEN vl.date >= '{filters.get('from_date') or ''}'
				ELSE 1=1
			END AND
			CASE
				WHEN IFNULL('{filters.get('to_date') or ''}', '') != ''
					THEN vl.date <= '{filters.get('to_date') or ''}'
				ELSE 1=1
			END AND
			CASE
				WHEN IFNULL({len(filters.get('purpose') or [])}, 0) != 0
					THEN vl.select_purpose in ({', '.join(f"'{i}'" for i in (filters.get('purpose') or ['', '']))})
				ELSE 1=1
			END
		GROUP BY vl.license_plate
	""", as_dict=True)

def get_columns(filters):
	return [
		{
			'fieldname': 'vehicle',
			'label': 'Vehicle',
			'fieldtype': 'Link',
			'options': 'Vehicle',
			'width': 200
		},
		{
			'fieldname': 'distance',
			'label': 'Distance Travelled',
			'fieldtype': 'Float',
			'width': 150
		}
	]